spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/MultipleJGitEnvironmentRepositoryFactory.java

```diff
@@ -38,23 +36,23 @@
 
 	private Optional<ConfigurableHttpConnectionFactory> connectionFactory;
 
-	private Optional<TransportConfigCallback> customTransportConfigCallback;
+	private final TransportConfigCallbackFactory transportConfigCallbackFactory;
 
 	@Deprecated
 	public MultipleJGitEnvironmentRepositoryFactory(ConfigurableEnvironment environment,
 			ConfigServerProperties server,
-			Optional<TransportConfigCallback> customTransportConfigCallback) {
-		this(environment, server, Optional.empty(), customTransportConfigCallback);
+			TransportConfigCallbackFactory transportConfigCallbackFactory) {
+		this(environment, server, Optional.empty(), transportConfigCallbackFactory);
 	}
 
 	public MultipleJGitEnvironmentRepositoryFactory(ConfigurableEnvironment environment,
```

Comment:
For backwards compatibility we should add a new constructor

Reply:
I moved the logic for creation of ssh transport config callbacks to `TransportConfigCallbackFactory`. If I have to preserve the `Optional<TransportConfigCallback>` constructor, either
- `MultipleJGitEnvironmentRepositoryFactory` would have to create `TransportConfigCallbackFactory`, which 1. is mixing concerns and 2. not possible, because `TransportConfigCallbackFactory` additionally has the `GoogleCloudSourceSupport` dependency.
- or: I would have to duplicate the ssh transport config callback creation code, which is 1. bad code duplication, 2. would lead to inconsistent behavior, because `GoogleCloudSourceSupport` would not be handled, when calling old `Optional<TransportConfigCallback>` constructor.
- or: `Optional<TransportConfigCallback>` constructor would only set the transport configuration callback to the `customTransportConfigCallback` (if it is present), ignoring ssh transport config callback and the google cloud source transport config callbacks, which would be - again - inconsistent to the `TransportConfigCallbackFactory` constructor.

My opinion is, that none of these solutions is desirable. But you are the one with more experience. Your call.

Reply:
Ryan? Which approach do you prefer?

Reply:
Why cant you just keep the current constructor marking it as `@Deprecated` and add a new one with the new parameter?

Reply:
Spring supports factories as beans rather than the actual bean. Invert control. Leave the constructor as is and create a factory bean rather than injecting the factory.

Reply:
@ryanjbaxter

> Why cant you just keep the current constructor marking it as @Deprecated and add a new one with the new parameter?

The logic before my change was:
 - if there is a bean of type `TransportConfigCallback` in Spring context, it will become the `customTransportConfigCallback`.
 - if there is a `customTransportConfigCallback`, use this bean as transport config callback.
 - otherwise: for all repositories, use `PropertiesBasedSshTransportConfigCallback` or `FileBasedSshTransportConfigCallback`, based on `MultipleJGitEnvironmentProperties.isIgnoreLocalSshSettings()`.

I have to add `GCSTransportConfigCallback` in to the mix without breaking backwards compatibility. Additionally you want me to keep the previous constructors, and add a third one.


**Approach 1**: not extracting transport callback construction logic into a separate class.

I add this constructor
```
public MultipleJGitEnvironmentRepositoryFactory(
    ConfigurableEnvironment environment,
    ConfigServerProperties server,
    Optional<ConfigurableHttpConnectionFactory> connectionFactory,
    Optional<TransportConfigCallback> customTransportConfigCallback,
    Optional<GoogleCloudSourceSupport> googleCloudSourceSupport)
```

Drawbacks:
- Since the other constructors do not receive an instance of `GoogleCloudSourceSupport`, `MultipleJGitEnvironmentRepositoryFactory` will work inconsistently, depending on which constructor you call. If you call one of the old constructors, Google Cloud Source repositories will just not work, even if all other conditions are met.
- `MultipleJGitEnvironmentRepositoryFactory` is polluted with details about repository vendors.
- The logic about choosing the proper transport config callback complicates `MultipleJGitEnvironmentRepositoryFactory`. But `MultipleJGitEnvironmentRepositoryFactory` should not be concerned with this.


**Approach 2**: extracting transport config construction logic into `TransportConfigCallbackFactory`.

I add this constructor
```
public MultipleJGitEnvironmentRepositoryFactory(
    ConfigurableEnvironment environment,
    ConfigServerProperties server,
    Optional<ConfigurableHttpConnectionFactory> connectionFactory,
    TransportConfigCallbackFactory transportConfigCallbackFactory) 
```
and change one of the old constructors to (the other delegates to this one):
```
@Deprecated
public MultipleJGitEnvironmentRepositoryFactory(ConfigurableEnvironment environment,
    ConfigServerProperties server,
    Optional<ConfigurableHttpConnectionFactory> connectionFactory,
    Optional<TransportConfigCallback> customTransportConfigCallback) {

  this(
      environment,
      server,
      connectionFactory,
      new TransportConfigCallbackFactory(customTransportConfigCallback.orElse(null), null)
  );
}
```


Some drawbacks of approach 1 are resolved, but others are added.

Drawbacks:
- `MultipleJGitEnvironmentRepositoryFactory` has to know the exact implementation of `TransportConfigCallbackFactory`, because it has to create an instance of it.
- this remains: Since the other constructors do not receive an instance of `GoogleCloudSourceSupport`, `MultipleJGitEnvironmentRepositoryFactory` will work inconsistently, depending on which constructor you call.


**Approach 3**: copy  transport config construction logic into `TransportConfigCallbackFactory`

I add this constructor
```
public MultipleJGitEnvironmentRepositoryFactory(
    ConfigurableEnvironment environment,
    ConfigServerProperties server,
    Optional<ConfigurableHttpConnectionFactory> connectionFactory,
    TransportConfigCallbackFactory transportConfigCallbackFactory) 
```
keep the other constructors as they are, and also keep the transport config creation logic in `MultipleJGitEnvironmentRepositoryFactory`. It will be used if one of the old constructors is called:

Drawbacks:
- I have to duplicate the code for the transport config creation logic.
- this remains: Since the other constructors do not receive an instance of `GoogleCloudSourceSupport`, `MultipleJGitEnvironmentRepositoryFactory` will work inconsistenly, depending on which constructor you call.


**Solution 4**: instantiate `GCSTransportConfigCallback` in `MultipleJGitEnvironmentRepositoryFactory` (as done for `FileBasedSshTransportConfigCallback` and `PropertiesBasedSshTransportConfigCallback`)

This is not a solution at all. Because dependency to Google library is optional, I cannot hard wire the Google Cloud Source implementation to `MultipleJGitEnvironmentRepositoryFactory`.


From my point of view there is no way to keep the old constructors **and** make `MultipleJGitEnvironmentRepositoryFactory` behave consistently, independent of the constructor you use. Am I missing something? Or do you prefer inconsistent behavior in favor of compile backwards compatibility?


@spencergibb

> create a factory bean rather than injecting the factory

I struggle to see how a factory bean will help me in this situation. 

I could create a factory bean, which will create a Google Cloud Source `TransportConfigCallback`. But this will mess up the legacy logic around `customTransportConfigCallback` (see above).

I could create a factory bean, which will create a `TransportConfigCallbackFactory`. This way `EnvironmentRepositoryConfiguration` will not have to instantiate `TransportConfigCallbackFactory` directly. But since I need a configuration class to tell Spring, that such a factory bean exists, I would just move the `new TransportConfigCallbackFactory()` code from one `@Bean` method to an other `@Bean` method. Not much is won. But on the other hand we well have a factory bean creating a factory, which is weird.

Am I missing your point, @spencergibb?